Native server hangs program after close if received at least one request

```console
$ deno --version
deno 1.9.2 (release, x86_64-apple-darwin)
v8 9.1.269.5
typescript 4.2.2
```

Using `oak@v7.4.0` seeing behaviour whereby a deno program for running an oak server will hang / fail to exit despite an abort signal being fired ( which closes the server ) _if_ at least one request has been received by the server in it's lifetime.

Small repro:

```console
deno run -A --unstable repro.ts
```

```ts
// repro.ts

import {
  Application,
  Router,
  RouterContext,
} from "https://deno.land/x/oak@v7.4.0/mod.ts";

const router = new Router();
const app = new Application();

router.get("/", (ctx: RouterContext) => {
  ctx.response.body = "hello";
});

app.use(router.routes());
app.use(router.allowedMethods());

const controller = new AbortController();
const { signal } = controller;

app.addEventListener("listen", async ({ hostname, port, secure }) => {
  const protocol = secure ? "https" : "http";
  const url = `${protocol}://${hostname}:${port}`;

  console.log(url); // http://localhost:8000

  // Omitting the following two lines results in expected behaviour,
  // i.e. the server closes after the abort signal is fired and the
  // deno programe terminates.
  const response = await fetch(url);
  console.log(await response.text()); // hello

  console.log("Will abort in 1 second...");

  setTimeout(() => {
    console.log("aborting...");
    controller.abort();
    console.log("aborted!");
  }, 1000);
});

await app.listen(
  { hostname: "localhost", port: 8000, signal },
);

console.log("Server closed, yet the program is still running at this point? 🤔");
console.log("Does seem to eventually stop after a minute or so...");
```

What is observed in the above example is shortly after the abort signal is fired, the bottom two logs are executed ( as expected ) but then the deno program hangs for a minute or so with no output - we would expect the program to terminate immediately.

Running similar examples against [superoak](https://github.com/asos-craigmorten/superoak) for testing reveals that there is potentially an unhandled op / resource that is keeping the program running, specifically we see errors such as...

```console
AssertionError: Test case is leaking async ops.
Before:
  - dispatched: 0
  - completed: 0
After:
  - dispatched: 10
  - completed: 9

Make sure to await all promises returned from Deno APIs before
finishing test case.
    at assert (deno:runtime/js/06_util.js:33:13)
    at asyncOpSanitizer (deno:runtime/js/40_testing.js:49:7)
    at async resourceSanitizer (deno:runtime/js/40_testing.js:73:7)
    at async Object.exitSanitizer [as fn] (deno:runtime/js/40_testing.js:100:9)
    at async TestRunner.[Symbol.asyncIterator] (deno:runtime/js/40_testing.js:272:13)
    at async Object.runTests (deno:runtime/js/40_testing.js:347:22)
    at async file:///Users/craigmorten/git/asos-craigmorten/superoak/$deno$test.ts:5:1
```

and if we set `sanitizeOps: false` then...

```console
AssertionError: Test case is leaking resources.
Before: {
  "0": "stdin",
  "1": "stdout",
  "2": "stderr"
}
After: {
  "0": "stdin",
  "1": "stdout",
  "2": "stderr",
  "7": "httpConnection"
}

Make sure to close all open resource handles returned from Deno APIs before
finishing test case.
    at assert (deno:runtime/js/06_util.js:33:13)
    at resourceSanitizer (deno:runtime/js/40_testing.js:84:7)
    at async Object.exitSanitizer [as fn] (deno:runtime/js/40_testing.js:100:9)
    at async TestRunner.[Symbol.asyncIterator] (deno:runtime/js/40_testing.js:272:13)
    at async Object.runTests (deno:runtime/js/40_testing.js:347:22)
    at async file:///Users/craigmorten/git/asos-craigmorten/superoak/$deno$test.ts:5:1
```

where the latter error is perhaps more revealing ( the prior could be an issue with the test / test framework )?

We are also still seeing issues running Oak's `closeServer.ts` example - 

1. Run the `closeServer.ts` example
2. Hit `localhost:8000/close`
3. Observe that the `Finished.` line is printed, but the deno program has not terminted
4. Hit `localhost:8000/close` or `localhost:8000`
5. Observe that an uncaught promise rejection error regarding the stream controller is thrown from Oak's `http_server_native.ts` code

```console
$ deno run -A --unstable ./examples/closeServer.ts 
Start listening on 127.0.0.1:8000
  using HTTP server: native
GET / - 1ms
GET /close - 0ms
Finished.
error: Uncaught (in promise) TypeError: The stream controller cannot close or enqueue.
            controller.close();
                       ^
    at ReadableStreamDefaultController.close (deno:op_crates/fetch/11_streams.js:3465:15)
    at serve (file:///Users/craigmorten/git/oakserver/oak/http_server_native.ts:179:24)
```

--- a/.github/workflows/oak-ci.yml
+++ b/.github/workflows/oak-ci.yml
@@ -26,7 +26,7 @@ jobs:
         run: deno lint --unstable
 
       - name: run tests
-        run: deno test --allow-read --allow-write
+        run: deno test --allow-read --allow-write --allow-net
 
       - name: run tests unstable
         run: deno test --coverage=./cov --allow-read --allow-write --allow-net --unstable

--- a/http_server_native.ts
+++ b/http_server_native.ts
@@ -173,6 +173,7 @@ export class HttpServerNative<AS extends State = Record<string, any>>
   #app: Application<AS>;
   #closed = false;
   #listener?: Deno.Listener;
+  #httpConnections: Set<Deno.HttpConn> = new Set();
   #options: Deno.ListenOptions | Deno.ListenTlsOptions;
 
   constructor(
@@ -198,16 +199,37 @@ export class HttpServerNative<AS extends State = Record<string, any>>
 
   close(): void {
     this.#closed = true;
+
     if (this.#listener) {
       this.#listener.close();
       this.#listener = undefined;
     }
+
+    for (const httpConn of this.#httpConnections) {
+      try {
+        httpConn.close();
+      } catch (error) {
+        if (!(error instanceof Deno.errors.BadResource)) {
+          throw error;
+        }
+      }
+    }
+
+    this.#httpConnections.clear();
   }
 
   listen(): Deno.Listener {
     return this.#listener = isListenTlsOptions(this.#options)
       ? Deno.listenTls(this.#options)
       : Deno.listen(this.#options);
+  }
+
+  #trackHttpConnection(httpConn: Deno.HttpConn): void {
+    this.#httpConnections.add(httpConn);
+  }
+
+  #untrackHttpConnection(httpConn: Deno.HttpConn): void {
+    this.#httpConnections.delete(httpConn);
   }
 
   [Symbol.asyncIterator](): AsyncIterableIterator<NativeRequest> {
@@ -218,19 +240,25 @@ export class HttpServerNative<AS extends State = Record<string, any>>
       const server = this;
       async function serve(conn: Deno.Conn) {
         const httpConn = serveHttp(conn);
+        server.#trackHttpConnection(httpConn);
+
         while (true) {
           try {
             const requestEvent = await httpConn.nextRequest();
+
             if (requestEvent === null) {
               return;
             }
+
             const nativeRequest = new NativeRequest(requestEvent, { conn });
             controller.enqueue(nativeRequest);
             await nativeRequest.donePromise;
           } catch (error) {
             server.app.dispatchEvent(new ErrorEvent("error", { error }));
           }
+
           if (server.closed) {
+            server.#untrackHttpConnection(httpConn);
             httpConn.close();
             controller.close();
           }

--- a/http_server_native_test.ts
+++ b/http_server_native_test.ts
@@ -2,7 +2,11 @@
 
 import { assertEquals, assertStrictEquals, unreachable } from "./test_deps.ts";
 
-import { hasNativeHttp, NativeRequest, HttpServerNative } from "./http_server_native.ts";
+import {
+  hasNativeHttp,
+  HttpServerNative,
+  NativeRequest,
+} from "./http_server_native.ts";
 
 import { Application } from "./application.ts";
 
@@ -51,14 +55,14 @@ test({
   name: "HttpServerNative closes gracefully after serving requests",
   async fn() {
     const app = new Application();
-    const listenOptions = { port: 4505 }
+    const listenOptions = { port: 4505 };
 
     const server = new HttpServerNative(app, listenOptions);
     server.listen();
 
     const expectedBody = "test-body";
 
-    (async() => {
+    (async () => {
       for await (const nativeRequest of server) {
         nativeRequest.respond(new Response(expectedBody));
       }
@@ -72,5 +76,5 @@ test({
     } finally {
       server.close();
     }
-  }
-})
+  },
+});

--- a/http_server_native_test.ts
+++ b/http_server_native_test.ts
@@ -1,8 +1,10 @@
 // Copyright 2018-2021 the oak authors. All rights reserved. MIT license.
 
-import { assertEquals, assertStrictEquals } from "./test_deps.ts";
+import { assertEquals, assertStrictEquals, unreachable } from "./test_deps.ts";
 
-import { hasNativeHttp, NativeRequest } from "./http_server_native.ts";
+import { hasNativeHttp, NativeRequest, HttpServerNative } from "./http_server_native.ts";
+
+import { Application } from "./application.ts";
 
 const { test } = Deno;
 
@@ -44,3 +46,31 @@ test({
     assertStrictEquals(await respondWithStack[0], response);
   },
 });
+
+test({
+  name: "HttpServerNative closes gracefully after serving requests",
+  async fn() {
+    const app = new Application();
+    const listenOptions = { port: 4505 }
+
+    const server = new HttpServerNative(app, listenOptions);
+    server.listen();
+
+    const expectedBody = "test-body";
+
+    (async() => {
+      for await (const nativeRequest of server) {
+        nativeRequest.respond(new Response(expectedBody));
+      }
+    })();
+
+    try {
+      const response = await fetch(`http://localhost:${listenOptions.port}`);
+      assertEquals(await response.text(), expectedBody);
+    } catch {
+      unreachable();
+    } finally {
+      server.close();
+    }
+  }
+})

--- a/test_deps.ts
+++ b/test_deps.ts
@@ -19,5 +19,6 @@ export {
   assertStrictEquals,
   assertThrows,
   assertThrowsAsync,
+  unreachable,
 } from "https://deno.land/std@0.105.0/testing/asserts.ts";
 export type { WebSocketEvent } from "https://deno.land/std@0.105.0/ws/mod.ts";

